test/cpp/jit/test_concat_opt.cpp

```diff
@@ -600,5 +600,129 @@ TEST(OptimizeConcatTest, NoOptimizationWhenInputListIsMutated) {
   }
 }
 
+TEST(OptimizeConcatTest, UseVariadicCat1) {
+  auto graph = std::make_shared<Graph>();
+
+  const std::string input =
+      R"IR(
+        graph(%0: Float(64, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu),
+              %1: Float(32, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu)):
+          %2 : int = prim::Constant[value=0]()
+          %input : Tensor[] = prim::ListConstruct(%0, %1)
+          %concat : Float(96, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu) = aten::cat(%input, %2)
+          return (%concat)
+      )IR";
+  parseIR(input, graph.get());
+  std::vector<at::Tensor> inputs = {
+      at::rand({64, 56, 56}, at::kCPU), at::rand({32, 56, 56}, at::kCPU)};
+  auto orig_outputs = runGraph(graph, inputs);
+
+  UseVariadicCat(graph);
+  graph->lint();
+  auto opt_outputs = runGraph(graph, inputs);
+
+  checkOutputs(orig_outputs, opt_outputs);
+
+  // After full concat optimization we should have the following graph:
+  //
+  //  graph(%0 : ...,
+  //        %1 : ...):
+  //    %zero : int = prim:Constant[value=0]()
+  //    %varcat : Tensor = prim::Concat(%0, %1, %zero)  // variadic cat
+  //    return (%varcat)
+  testing::FileCheck()
+      .check_count("= prim::Concat(", 1, /*exactly*/ true)
+      ->check_count("= aten::cat(", 0, /*exactly*/ true)
+      ->check_count("= prim::ListConstruct(", 0, /*exactly*/ true)
+      ->run(*graph);
+}
+
+TEST(OptimizeConcatTest, UseVariadicCat2) {
```

Comment:
The second test case looks just like the first one except for more elements in the list. Are they exercising different logics in the implementation? If not, perhaps keeping UseVariadicCat2 would be sufficient.

Maybe we should add a test case where the list constructed is not only used in concat but also used in another computation. The test will check that the constructed list is not preserved.

```
    R"IR(        graph(%0: Float(64, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu), 
             %1: Float(32, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu)): 
             %2 : int = prim::Constant[value=0]()
             %input : Tensor[] = prim::ListConstruct(%0, %1) 
            %concat : Float(96, 56, 56, strides=[3136, 56, 1], requires_grad=0, device=cpu) = aten::cat(%input, %2)
            return (%concat,%input)      )IR";
```

Reply:
Sure, added that test case. Retained only one for the positive case.

Note that, with the case of List output being used in multiple places, the transformation would not happen now, since we are explicitly checking for > 1 use of the List.